col/src/main/java/vct/col/ast/util/NameScanner.scala

```diff
@@ -0,0 +1,210 @@
+    }
+  }
+
+  private def setWrite(name: String): Unit =
+    freeNames.get(name) match {
+      case Some(value) => value.writtenTo = true
+      case None => Abort("Cannot set write for non-existent variable %s", name)
+    }
+
+  private def push(): Unit = frameStack.push(mutable.Set())
+
+  private def pop(): Unit = frameStack.pop()
+
+  /**
+    * Returns free names as a java map with insertion order retained
+    */
+  def freeNamesJava: util.Map[String, Type] = {
+    val map = new util.LinkedHashMap[String, Type]()
+    freeNames.foreach(mapEntry => map.put(mapEntry._1, mapEntry._2.typ))
+    map
+  }
+
+  /**
+    * All variable names that are read from or written to
+    * @return
+    */
+  def accesses: Set[String] = freeNames.keySet.toSet
+
+  def writes: Set[String] = freeNames.filter(p => p._2.writtenTo).keySet.toSet
+
+  def reads: Set[String] = freeNames.filter(p => !p._2.writtenTo).keySet.toSet
+
+  override def visit(e: NameExpression): Unit = e.getKind match {
+    case Reserved =>
+    case Label =>
+    case Field | Local | Argument =>
+      val name = e.getName
+      val t = e.getType
+
+      freeNames.get(name) match {
+        case Some(value) =>
+          // If the name is already a free variable, the types must match
+          if (t != null && !(value.typ == t)) {
+            Fail("type mismatch %s != %s", t, value.typ);
+          }
+        case None if !hasDecl(name) =>
+          // If it is new and there is no decl, declare a new free variable
+          put(name, t)
+        case _ =>
+          // Otherwise ignore it
+      }
+    case Unresolved =>
+      e.getName match {
+        case "tcount" =>
+        case "gsize" =>
+        case "tid" =>
+        case "gid" =>
+        case "lid" =>
+        case "threadIdx" =>
+        case "blockIdx" =>
+        case "blockDim" =>
+          put(e.getName, e.getType)
+      }
+    case _ =>
+      Abort("missing case %s %s in name scanner", e.getKind, e.getName)
+  }
+
+  override def visit(assignment: AssignmentStatement): Unit = {
+    super.visit(assignment)
+
+    assignment.location match {
+      case name: NameExpression if !hasDecl(name.name) => setWrite(name.name)
+      case _ =>
+    }
+  }
+
+  override def visit(d: DeclarationStatement): Unit = {
+    // Discovered a new declaration
+    frameStack.top.add(d)
+
+    super.visit(d)
+  }
+
+  override def visit(s: LoopStatement): Unit = {
+    push()
+
+    super.visit(s)
+
+    pop()
+  }
+
+  override def visit(e: BindingExpression): Unit = if (e.getDeclCount == 1) {
+    push()
+
+    super.visit(e)
+
+    pop()
+  }
+  else Abort("missing case in free variable detection")
+
+  override def visit(s: ForEachLoop): Unit = {
+    push()
+
+    super.visit(s)
+    if (s.get_before != null) s.get_before.accept(this)
+    if (s.get_after != null) s.get_after.accept(this)
+    // Ensure before/after are not scanned after pop()
+    auto_before_after = false
```

Comment:
This seems kinda leaky

Reply:
Yeah, and I'm not sure why ForEachLoop does get to scan before/after, and not ParallelBlock and MethodInvokation. (the auto_before_after is meant to be used this way, so that part is okay I think as long as the flag itself exists)

Reply:
Even more so: I'd argue that with/then blocks _should_ be scanned, because if they use free variables those should be taken into account as well. Only the target variable names within given/yields should be ignored.

Reply:
The reason why it is this way (I think) is that if there is a decl in the before block, it should only be in scope for the for loop and the after block. Re: par blocks, they don't have before/after, so I removed it there. Re: method invokations and with/then, there I will additionally implement that they are incorporated as well, taking into account that given/yields variables should not be recorded as free variables.